qep-14-authentication-system.rst

```diff
@@ -0,0 +1,503 @@
+This might mean limiting cross-plugin communication as well, but maybe only
+between third-party plugins.
+
+Another good solution is to issue code-signing certificates to vetted plugin
+authors. Then validate the plugin's certificate upon loading. If need be the
+user can also directly set an untrusted policy for the certificate associated
+with the plugin using existing certificate management dialogs.
+
+Alternatively, access to sensitive auth system data from Python could *never* be
+allowed, and only the use of QGIS core widgets, or duplicating auth system
+integrations, would allow the plugin to work with resources that have an
+``authcfg``, while keeping master password and auth config loading in the realm
+of the main app.
+
+The same security concerns apply to C++ plugins, though it will be harder to
+restrict access, since there is no function binding to simply be removed as with
+Python.
+
+6. Further Improvements
+-----------------------
+
+General auth system improvements to be considered (no particular order):
+
+* Have security guru/firm audit implementation (I'm no guru)
+* Integrate auth system with more database connection configs
+* Integrate auth system with Plugin Manager connections
+* Integrate auth system with all HTTP connection classes
+* Integrate auth system into QGIS Server, where user prompts are not supported
+* Integrate master password entry with platform-specific password managers, so
+  user does not need to enter it and automated pre-population scripts are easier
+  to code
+* Finish implementation of auth config 'resource' (auto-auth via matched
+  resource URI)
+* Try moving auth system integration code (when authcfg has been assigned) into
+  ``QgsNetworkAccessManager`` instead of within individual data providers and
+  pass authcfg to network manager
+* Warn user when secure parts of auth system are accessed by Python (see
+  above)
+* PyQGIS plugins may need sandboxed to protect against, and selectively allow,
+  access to auth configs
+* Migrate all 'core' PyQGIS plugins that need to manage their own connections to
+  using core auth system calls and/or core connections (which would auto-manage
+  auth system calls), or isolate them from third-party plugins, if possible
+* Switch master password memory allocation from QString (insecure) to
+  ``QCA::SecureArray``
+* Add conversion button, to convert existing plain auth to auth config
+* Add optional ability to edit the auth id for a configuration (user must
+  confirm before operation)
+* Add ability to change, edit or remove authcfg from existing layer in
+  properties dialog
+* Add simple read-only text field in layer properties dialog to quickly copy
+  authcfg
+* Add copy/paste/add/edit/remove of authcfg in layer contextual menu in Legend
+  panel
+* On layer load, notify user if any associated authcfg is missing in auth
+  database
+* Add layer authcfg (re)assignment in Handle Bad Layers dialog (can currently
```

Comment:
Good to see this listed - I think these two tasks are important and should be prioritised  if possible.